packages/rule-components/src/RuleTable/RuleTable.js

```diff
@@ -0,0 +1,249 @@
+        const { filterValues } = this.state;
+        fetchData({
+            meta: calculateMeta(rules.meta),
+            filterValues,
+            sortBy: { index, direction, column: columns[index - Boolean(detail)] }
+        });
+    }
+
+    onPaginate = (perPage, page) => {
+        const { fetchData, rules, columns, detail, sortBy } = this.props;
+        const { filterValues } = this.state;
+        fetchData({
+            meta: { ...calculateMeta(rules.meta), perPage, page },
+            filterValues,
+            sortBy: { ...sortBy, column: columns[sortBy.index - Boolean(detail)] }
+        });
+    }
+
+    render() {
+        const {
+            rules,
+            filters,
+            columns,
+            detail,
+            actions,
+            ariaLabel,
+            fetchData,
+            isLoading,
+            sortBy,
+            toolbarProps,
+            filterValues: _filterValues,
+            ...props
+        } = this.props;
+        const { expanded, filterValues } = this.state;
+        const { meta, data } = rules;
+
+        const defaultMeta = calculateMeta(meta);
+        const pagination = {
+            ...defaultMeta,
+            onPerPageSelect: (_e, perPage) => this.onPaginate(perPage, 1),
+            onSetPage: (_e, page) => this.onPaginate(defaultMeta.perPage, page)
+        };
+        const filterItems = Object.entries(filters).map(([ key, filter ]) => ({
+            ...filter({
+                onChange: (...props) => this.onFilterChange(key, ...props),
+                value: filterValues[key],
+                filterKey: key
+            })
+        }));
+
+        return <section className="rhcs-c-rules-table">
+            <PrimaryToolbar
+                { ...filterItems && filterItems.length > 0 && { filterConfig: { items: filterItems } }}
+                activeFiltersConfig={{
+                    filters: Object.entries(filterValues).map(([ key, value ]) => calculateActiveFilters(
+                        filterItems.find(({ filterKey }) => filterKey === key),
+                        value,
+                        key
+                    )).filter(Boolean),
+                    onDelete: this.onFilterDelete
+                }}
+                pagination={pagination}
+                { ...toolbarProps }
+            />
+            <Table
+                {...props}
+                {...sortBy && { onSort: this.onSort }}
+                actions={!isLoading && actions}
+                aria-label={ariaLabel}
+                cells={columns.map(({ transforms, ...column }) => ({
+                    ...column,
+                    transforms: [
+                        ...transforms || [],
+                        ...!isLoading && sortBy && !column.disableSort ? [ sortable ] : []
+                    ]
+                }))}
+                rows={
+                    isLoading ?
+                        [ ...new Array(5) ].map(() => ({
```

Comment:
shouldn't this constant be moved to top of the file or something?

Reply:
Not really, if the data are not loading this is not invoked at all. If it was moved to the top of component it would mean that whenever component is used it will be invoked.

Reply:
What do you mean? I'm saying that having hardcoded constants is a bad practice.

Reply:
I suggest to use here method `fill`: `Array(5).fill()`. It will be more readable and you don't need to copy array.

Reply:
For me it's easier to read spread than fill, sorry.

I've added new prop so the number of loading bars/rows can be changed.

Reply:
@karelhala, great, thanks